addons/fetchmail/models/fetchmail.py

```diff
@@ -108,22 +109,26 @@ def set_draft(self):
     def connect(self):
         self.ensure_one()
         if self.type == 'imap':
-            if self.is_ssl:
-                connection = IMAP4_SSL(self.server, int(self.port))
-            else:
-                connection = IMAP4(self.server, int(self.port))
+            try:
+                # HACK https://bugs.python.org/issue38615
```

Comment:
We should be able to find a thread-safe way of setting this timeout.

It seems `IMAP4.open()` is [stable and intended to be overridable](https://github.com/python/cpython/blob/3.6/Lib/imaplib.py#L281-L300), all the way from 2.7 to 3.8.
And if we ignore P2 and only target 12.0 or 13.0, `IMAP4._create_socket()` looks like an even better target.

So a monkey-patch along those lines might work (at the cost of some sys.audit event in 3.8, whatever)

```python
# imaplib has no default timeout - monkeypatch it to support one
def _create_socket_timeout(self):
    host = None if not self.host else self.host
    return socket.create_connection((host, self.port), MAIL_TIMEOUT)
IMAP4._create_socket = _create_socket_timeout
```

Reply:
If our purpose is to do it thread-safe, monkey-patching a method in the standard library doesn't seem the best way.

Wouldn't it be better to use a subclass that overrides that method?

Reply:
> Wouldn't it be better to use a subclass that overrides that method?

Issue's them you have to override both IMAP4 and IMAP4_SSL, because the latter calls `IMAP4._create_socket` directly rather than call `self._create_socket` or whatever: https://github.com/python/cpython/blob/86c17c06c9420040c79c29ecf924741f37975342/Lib/imaplib.py#L1290-L1293

Reply:
> If our purpose is to do it thread-safe, monkey-patching a method in the standard library doesn't seem the best way.

Having IMAP4 default to using a timeout also makes sense in the case of an Odoo process - it's a rather specific use case, so the global effect is intended. Who wants an IMAP connection with no timeout at all?

It seems much harder to predict what side effects a default socket timeout could have on other threads.

Reply:
@xmo-odoo Yes but if I subclass both `IMAP4` and `IMAP4_SSL` classes, they'll call what I tell them to call, right? So that wouldn't be a problem at all.

Reply:
> @xmo-odoo Yes but if I subclass both IMAP4 and IMAP4_SSL classes, they'll call what I tell them to call, right? So that wouldn't be a problem at all.

Sure but you have to copy/paste/modify IMAP4_SSL entirely, and as @odony note that means any user of the regular IMAP classes gets no timeout, so it's more work for less safety.

Reply:
Fixed. I ended up using `mock`, which is purpose-built for handling monkey patches. The patch looks better this way IMHO.